pyccel/parser/syntax/openmp.py

```diff
@@ -11,396 +11,962 @@
+    def __init__(self, **kwargs):
+        super().__init__(OmpAnnotatedComment, _valid_atomic_clauses, **kwargs)
 
 class OmpEndClause(BasicStmt):
-    """Class representing a ."""
+    """Class representing an end construct."""
     def __init__(self, **kwargs):
-        """
-        """
         self.construct = kwargs.pop('construct')
         self.simd      = kwargs.pop('simd', '')
         self.nowait    = kwargs.pop('nowait', '')
 
-        super(OmpEndClause, self).__init__(**kwargs)
+        construct = ' '.join(self.construct)
+        txt = 'end {0} {1} {2}'.format(construct, self.simd, self.nowait)
+
+        self._expr = Omp_End_Clause(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        return self._expr
+
+class OmpFinal(BasicStmt):
+    """Class representing a final clause"""
+    def __init__(self, **kwargs):
+        self.final = kwargs.pop('final')
+
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
-            print("> OmpEndClause: expr")
+            print("> {}: expr".format(type(self).__name__))
 
-        txt = 'end {0} {1} {2}'.format(self.construct, self.simd, self.nowait)
-        return Omp_End_Clause(txt)
+        final = self.final
+        return 'final({})'.format(final)
 
 class OmpNumThread(BasicStmt):
-    """Class representing a ."""
+    """Class representing a num_thread clause."""
     def __init__(self, **kwargs):
-        """
-        """
         self.thread = kwargs.pop('thread')
 
-        super(OmpNumThread, self).__init__(**kwargs)
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         # TODO check if variable exist in namespace
         if DEBUG:
-            print("> OmpNumThread: expr")
+            print("> {}: expr".format(type(self).__name__))
 
         thread = self.thread
         return 'num_threads({})'.format(thread)
 
+class OmpNumTeams(BasicStmt):
+    """Class representing a num_teams clause."""
+    def __init__(self, **kwargs):
+        self.teams = kwargs.pop('teams')
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        # TODO check if variable exist in namespace
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        teams = self.teams
+        return 'num_teams({})'.format(teams)
+
+class OmpThreadLimit(BasicStmt):
+    """Class representing a thread_limit clause."""
+    def __init__(self, **kwargs):
+        self.limit = kwargs.pop('limit')
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        # TODO check if variable exist in namespace
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        limit = self.limit
+        return 'thread_limit({})'.format(limit)
+
+class OmpNumTasks(BasicStmt):
+    """Class representing a num_tasks clause."""
```

Comment:
num_tasks -> NumTasks.

Reply:
there is no `NumTasks ` in openmp, this class is representing a `num_tasks` clause